### ISSUE TITLE
self included in list of params for method

### ISSUE THREAD

**Issue by thequackdaddy** [numpy/numpydoc#220]
Hello

I'm trying to fix some issues with statsmodels documentation. 

One issue that we've uncovered is that numpydoc 0.9.0 (and 0.9.1) includes `self` in the list of params for a method, when it would be preferable if it didn't. That's different than what 0.8.0 did. 

Screenshot of 0.9.0...

<img width="825" alt="Screen Shot 2019-05-27 at 12 19 05 PM" src="https://user-images.githubusercontent.com/8186432/58432945-88bd8d00-807a-11e9-8f56-1f40c46b5a76.png">

Screenshot of 0.8.0...

<img width="833" alt="Screen Shot 2019-05-27 at 12 23 40 PM" src="https://user-images.githubusercontent.com/8186432/58432956-8f4c0480-807a-11e9-8bba-83e1900c0f68.png">

Any ideas how to fix?

**Comment by jnothman** [comment]
Oh, that's mysterious!

**Comment by pv** [comment]
Is it numpydoc, or sphinx version upgrade?

**Comment by thequackdaddy** [comment]
Its just a numpydoc upgrade. I am using sphinx 2.0.1 on both. 

From what I I can figure, you can do...

```python
from numpydoc.docscrape_sphinx import get_doc_object
import statsmodels.api as sm

get_doc_objecdt(sm.GLM)

```

and you get the same output for both. That's just an autosummray table with just the method names (no signature/parameters). 

So that looks right. But the signature is different. I can't really tell how the signature is generated from the code.

**Comment by thequackdaddy** [comment]
Git bisect has lead me to think the issue is here:

https://github.com/numpy/numpydoc/commit/2cdd4d9e10ffdd1cf5f2359e01e3e8ef59d42c88

**Comment by jnothman** [comment]
I can't believe that. Could it be the parent of that commit,
https://github.com/numpy/numpydoc/commit/e47b9404963ad2a75a11d167416038275c50d1c5
??

**Comment by thequackdaddy** [comment]
I'm pretty certain that is it. I changed that code back in a development version and it works. 

I can't seem to figure out how the code works, however... 

```python
In [1]: import statsmodels.api as sm

In [2]: from numpydoc.docscrape_sphinx import SphinxDocString, get_doc_object

In [3]: import pydoc

In [4]: SphinxDocString(pydoc.getdoc(sm.GLM.fit))
Out[4]: <numpydoc.docscrape_sphinx.SphinxDocString at 0x1204bf630>

In [5]: SphinxDocString(pydoc.getdoc(sm.GLM.fit))['Signature']
Out[5]: ''

In [6]: get_doc_object(sm.GLM.fit)['Signature']
Out[6]: "fit(self, start_params=None, maxiter=100, method='IRLS', tol=1e-08, scale=None, cov_type='nonrobust', cov_kwds=None, use_t=None, full_output=True, disp=False, max_start_irls=3, \\*\\*kwargs)"

``` 

I don't know why running this way, the "old" way doesn't work. It must be adding the signature *before*  it goes through this. I'm not sure how the `app.connect('autodoc-process-signature', mangle_signature)` line works...

**Comment by thequackdaddy** [comment]
(my guess is that autodoc is parsing it, and realizing that this is method and dropping the first parameter, then prepending the signature on the docs. Not sure how to setup a test of that, however.)

**Comment by thequackdaddy** [comment]
Ok, better guess. I think before, this was broken and didn't return signatures when they weren't explicitly stated in the docstring. Then I'm guessing sphinx added the correct one. 

With the "fix" (that reverts behavior) `doc['Signature']` is `''`. 
```
loading intersphinx inventory from https://docs.scipy.org/doc/numpy/objects.inv...
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://matthew-brett.github.io/pydagogue/objects.inv...
loading intersphinx inventory from https://patsy.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from https://pandas.pydata.org/pandas-docs/stable/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 90 source files that are out of date
updating environment: 376 added, 0 changed, 0 removed
> /Users/peter/git/numpydoc/numpydoc/numpydoc.py(222)mangle_signature()ar_model.GLM
-> sig = doc['Signature'] or getattr(obj, '__text_signature__', None)
(Pdb) doc
<numpydoc.docscrape_sphinx.SphinxDocString object at 0x12483ae48>
(Pdb) doc['Signature']
''
(Pdb) c
> /Users/peter/git/numpydoc/numpydoc/numpydoc.py(222)mangle_signature()ar_model.GLM.fit
-> sig = doc['Signature'] or getattr(obj, '__text_signature__', None)
(Pdb) doc
<numpydoc.docscrape_sphinx.SphinxDocString object at 0x126d64320>
(Pdb) doc['Signature']
''
(Pdb)
```

**Comment by thequackdaddy** [comment]
Ok I think it boils down to this... 

Before, this whole section got skipped... its not in the inheritance of `SphinxDocString`. 

https://github.com/numpy/numpydoc/blob/2fa5256d35cb75c5032e486a543271a420b6971c/numpydoc/docscrape.py#L564-L590

Now, this runs, and adds the signature that didn't exist before. Because the signature is added, my theory is that autodoc doesn't use its (slightly different) logic to add the signature. 

https://github.com/sphinx-doc/sphinx/blob/f63abac2cad2664a8af816017f0f997bae510d14/sphinx/ext/autodoc/__init__.py#L1313-L1324

**Comment by jnothman** [comment]
What a strange interaction. I'm struggling to get my head around it.

**Comment by jnothman** [comment]
I don't understand how #200 worked really.

**Comment by thequackdaddy** [comment]
Let me rephrase. 

It was broken before. `SphinxDocString(pydoc.getdoc(class.method))['Signature']` would always return `''` when the signature is not included in the actual doctoring. 

#200 "fixed" it so that.

Now `get_doc_object(class.method)` creates an instance of `SphinxFunctionDoc(obj)`. That inherits `FuncitonDoc` and runs `FunctionDoc.__init__` where the code above generates a signature. So `get_doc_object(class.method)['Signature']` will not be `''`. There is no logic to eliminate `self` from methods in this code. 

 Then--and here I haven't looked at sphinx's autodoc--since a signature already exists, autodoc just relies on the provided signature.  

Before #200, since there was no signature, sphinx's auto doc used its logic to create the signature and its logic was clever enough to drop `self`.

**Comment by jnothman** [comment]
so you think the solution is to adopt autodoc's logic?

**Comment by thequackdaddy** [comment]
> so you think the solution is to adopt autodoc's logic?

I guess that would make sense. Either (hard way) duplicate their logic or (easy way) create a Signature only if it already exists in the docstring. Otherwise, just create a Signature of `''`.

**Comment by thequackdaddy** [comment]
I updated #225 to add the functionality I'm thinking about. Let me know if you think this is on the right path.

**Comment by fladd** [comment]
Is there a way to fix this for ourselves (besides reverting to 0.8.0)?

**Comment by thequackdaddy** [comment]
Closed by #221
